internal/metadataproviders/aws/eks/metadata.go

```diff
@@ -0,0 +1,175 @@
+	ClusterVersion() (string, error)
+	GetK8sInstanceMetadata(ctx context.Context) (InstanceMetadata, error)
+	GetInstanceMetadata(ctx context.Context) (InstanceMetadata, error)
+	SetRegionInstanceID(region, instanceID string)
+}
+
+type ec2Client interface {
+	DescribeInstances(ctx context.Context, params *ec2.DescribeInstancesInput, optFns ...func(*ec2.Options)) (*ec2.DescribeInstancesOutput, error)
+}
+
+type metadataClient struct {
+	instanceMetadata InstanceMetadata
+	clientset        kubernetes.Interface
+	ec2Client        ec2Client
+	nodeName         string
+}
+
+type InstanceMetadata struct {
+	AvailabilityZone string
+	AccountID        string
+	InstanceID       string
+	Region           string
+	ClusterName      string
+	ImageID          string
+	InstanceType     string
+	Hostname         string
+}
+
+var _ Provider = (*metadataClient)(nil)
+
+func NewProvider(cfg aws.Config, nodeName string) (Provider, error) {
+	conf, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize provider. Could not load cluster config: %w", err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(conf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize provider. Could not create clientset for Kubernetes client: %w", err)
+	}
+
+	return &metadataClient{
+		instanceMetadata: InstanceMetadata{},
+		nodeName:         nodeName,
+		clientset:        clientset,
+		ec2Client:        ec2.NewFromConfig(cfg),
+	}, nil
+}
+
+func (c *metadataClient) ClusterVersion() (string, error) {
+	serverVersion, err := c.clientset.Discovery().ServerVersion()
+	if err != nil {
+		return "", fmt.Errorf("failed to retrieve cluster version: %w", err)
+	}
+	return serverVersion.GitVersion, nil
+}
+
+// GetK8sInstanceMetadata retrieves region, instanceID, and availabilityZone attributes from the K8s node's providerID.
+// It requires the node name which is passed to the constructor.
+// If region or instanceID are not found, it returns an error as they are mandatory to query ec2 API for the full metadata.
+func (c *metadataClient) GetK8sInstanceMetadata(ctx context.Context) (InstanceMetadata, error) {
+	if c.nodeName == "" {
+		return InstanceMetadata{}, errors.New("can't get K8s Instance Metadata; node name is empty")
+	}
+
+	node, err := c.clientset.CoreV1().Nodes().Get(ctx, c.nodeName, metav1.GetOptions{})
+	if err != nil {
+		return InstanceMetadata{}, fmt.Errorf("can't get K8s Instance Metadata; failed to retrieve k8s node %w", err)
+	}
+
+	region, availabilityZone, instanceID := parseRegionAndInstanceID(node.Spec.ProviderID)
+	if region == "" || instanceID == "" {
+		return InstanceMetadata{}, fmt.Errorf("failed to retrieve region or instanceId from mockProviderID: %s", node.Spec.ProviderID)
+	}
+
+	c.instanceMetadata = InstanceMetadata{Region: region, InstanceID: instanceID, AvailabilityZone: availabilityZone}
+	return c.instanceMetadata, nil
+}
+
+func parseRegionAndInstanceID(input string) (string, string, string) {
```

Comment:
Should we maybe use regex here instead? Something like `^aws:///(?P<az>(?P<region>\w+-\w+-\d+)\w*)/(?P<instance>i-[0-9a-fA-F]{8,})$`. Don't trust this though

Reply:
I went with split as the delimiter is well defined and does not clash/get used in the targeted strings. 
Is there a specific reason to use regex? Especially that I added a safety guard for NPEs.
Thanks!

Reply:
Added a check for `aws:` split